In [1]:
import numpy as np
import matplotlib.pyplot as plt
import numpy.linalg as lin

***Exercise 1:*** Theory: This part requires no implementation, you are allowed to write your answers on a different document (pdf, png, ...) if you are not confortable with markdown.

Consider the matrix
$$A = \left(\begin{array}{ccc}
1 & 0 & 1 \\ 0 & 1 & 1 \\ -1 & 1 & 1
\end{array}\right).$$

**a)** Compute its characteristic polynomial.

**Answer:**  

We compute that 
\begin{align} 
    \det{\left( A - \lambda I_3 \right)} 
    &= \begin{vmatrix} 
    1 - \lambda & 0 & 1 \\ 
    0 & 1 - \lambda & 1 \\ 
    -1 & 1 & 1 - \lambda 
    \end{vmatrix} 
    \\ 
    &= (1 - \lambda) \begin{vmatrix} 
        1 - \lambda & 1 \\ 
        1 & 1 - \lambda 
    \end{vmatrix} - 1 \begin{vmatrix} 
        0 & 1 \\ 
        1 - \lambda & 1 
    \end{vmatrix} 
    \\ 
    &= (1 - \lambda) \left( (1 - \lambda)^2 - 1 \right) - \left( 0 - (1 - \lambda) \right) 
    \\ 
    &= (1 - \lambda)^3 - (1 - \lambda) + (1 - \lambda) 
    \\ 
    &= (1 - \lambda)^3 
\end{align} 

Therefore, its characteristic polynomial is 
$$ (1 - \lambda)^3 = 0 $$ 

**b)** Find the eigenvalue(s) of $A$ and the associated algebraic multiplicitie(s)?

**Answer:** 

The eigenvalue of $A$ is $\lambda = 1$ with algebraic multiplicity $m = 3$. 

**c)** Find all its eigenvectors and deduce the geometric multiplicitie(s) of the eigenvalue(s). 

**Answer:** 

We compute the eigenvector $v$ of $\lambda = 1$. 
$$ 
Av = \lambda v 
\quad \Longrightarrow \quad 
\begin{pmatrix} 
    1 & 0 & 1 \\ 0 & 1 & 1 \\ -1 & 1 & 1 
\end{pmatrix} 
\begin{pmatrix} 
    v_1 \\ v_2 \\ v_3 
\end{pmatrix} 
= \begin{pmatrix} 
    v_1 \\ v_2 \\ v_3 
\end{pmatrix} 
\quad \Longrightarrow \quad 
\begin{cases} 
    v_1 = v_2 \\ 
    v_3 = 0 
\end{cases} 
\quad \Longrightarrow \quad 
v = \begin{pmatrix} 
    1 \\ 1 \\ 0 
\end{pmatrix} 
$$ 

Hence, we can deduce that 
$$ 
\text{ker} \left( \lambda I_3 - A \right) = 
\text{Span} \begin{pmatrix} 
    1 \\ 1 \\ 0 
\end{pmatrix} 
\quad \Longrightarrow \quad 
\mu = \dim \text{ker} \left( \lambda I_3 - A \right) = 1 
$$ 

It means that there is only one linearly independent eigenvector associated to $\lambda = 1$. Therefore, we can conclude that its geometric multiplicity is $\mu = 1$. 

---
---

**Exercise 2:** Implementation:

Consider the matrix 
$$B_\epsilon = \frac{1}{2}\left( \begin{array}{cc} 
\frac{1}{\epsilon} + \epsilon & \frac{1}{\epsilon} - \epsilon \\  \frac{1}{\epsilon} - \epsilon & \frac{1}{\epsilon} + \epsilon \end{array}\right),$$
such that 
$$B_\epsilon = PDiag\left(\epsilon,\frac{1}{\epsilon}\right)P^{-1} \quad \text{with}\quad P = \left(\begin{array}{cc} 1 & 1 \\ -1 & 1\end{array}\right),$$ 
and 
$$B_\epsilon^{-1} = \frac{1}{2}\left( \begin{array}{cc} 
\epsilon+\frac{1}{\epsilon} & \epsilon-\frac{1}{\epsilon} \\ \epsilon-\frac{1}{\epsilon}& \epsilon+\frac{1}{\epsilon} \end{array}\right) = P Diag\left(\frac{1}{\epsilon}, \epsilon\right) P^{-1}.$$

**a)** Implement Gauss-Seidel algorithm for solving systems of 2 linear equations with 2 unknowns, i.e. $BU = c$ with $B\in\mathbb{R}^{2\times 2}$ and $c\in\mathbb{R}^2$. Test it with the given data.

*Hint: Jacobi algorithm is given as an example. You may copy-paste it and modify it to obtain Gauss-Seidel algorithm.*

In [2]:
def Jacobi(B, c, Uinit, kmax, TOL):
    # solve the equation BU = c using the Jacobi method
    if (len(c) != 2 or len(B[0,:]) != 2 or len(B[:,0]) != 2): 
        return "size error" 
    
    U_new   = np.copy(Uinit) 
    epsilon = lin.norm(np.matmul(B,U_new) - c) 
    
    k       = 0 
    while (k < kmax) and (epsilon > TOL): 
        k +=1 
        U = np.copy(U_new) 
        
        U_new[0] = (c[0] - B[0,1] * U[1]) / B[0,0] 
        U_new[1] = (c[1] - B[1,0] * U[0]) / B[1,1] 
        
        epsilon = lin.norm(np.matmul(B,U_new) - c) 
        
    return U 

In [3]:
eps    = 0.01
Beps   = .5 * np.array([[eps+1/eps,eps-1/eps],[eps-1/eps,eps+1/eps]])
c      = np.array([1.,2.])
U_init = np.ones(2)
k_max  = 10**5 
TOL    = .00001

U_sol = Jacobi(Beps, c, U_init, k_max, TOL)
print("Matrix\n", Beps,"\n")
print("Numerical solution\n", U_sol,"\n")
print("Residual\n",np.matmul(Beps,U_sol)-c)

Matrix
 [[ 50.005 -49.995]
 [-49.995  50.005]] 

Numerical solution
 [149.99432963 150.00432959] 

Residual
 [-4.45432397e-06 -8.95343783e-06]


In [4]:
def GS(B, c, Uinit, kmax, TOL):
    # solve the equation BU = c using the Gauss-Seidel algorithm
    if (len(c) != 2 or len(B[0,:]) != 2 or len(B[:,0]) != 2): 
        return "size error" 
    
    U_new   = np.copy(Uinit) 
    epsilon = lin.norm(np.matmul(B,U_new) - c) 
    
    k       = 0 
    while (k < kmax) and (epsilon > TOL): 
        k +=1 
        U = np.copy(U_new) 
        
        U_new[0] = (c[0] - B[0,1] * U[1]) / B[0,0] 
        U_new[1] = (c[1] - B[1,0] * U_new[0]) / B[1,1] 
        
        epsilon = lin.norm(np.matmul(B,U_new) - c) 
        
    return U 

In [5]:
eps    = 0.01
Beps   = .5* np.array([[eps+1/eps,eps-1/eps],[eps-1/eps,eps+1/eps]])
c      = np.array([1.,2.])
U_init = np.ones(2)
k_max  = 10**5
TOL    = .00001

U_sol = GS(Beps, c, U_init, k_max, TOL)
print("Matrix\n", Beps,"\n")
print("Numerical solution\n", U_sol,"\n")
print("Residual\n",np.matmul(Beps,U_sol)-c)

Matrix
 [[ 50.005 -49.995]
 [-49.995  50.005]] 

Numerical solution
 [149.99449986 150.00449996] 

Residual
 [-1.00018224e-05  0.00000000e+00]


**b)** Implement the alternating Gauss-Seidel algorithm. Test it with the same parameters. 

In [6]:
def GS_altern(B, c, Uinit, kmax, TOL):
    # solve the equation BU = c using the alternating Gauss-Seidel algorithm
    if (len(c) != 2 or len(B[0,:]) != 2 or len(B[:,0]) != 2): 
        return "size error" 
    
    U_new   = np.copy(Uinit) 
    epsilon = lin.norm(np.matmul(B,U_new) - c) 
    
    k       = 0 
    while (k < kmax) and (epsilon > TOL): 
        k +=1 
        U = np.copy(U_new) 
        
        if (k % 2 == 0):  
            U_new[1] = (c[1] - B[1,0] * U[0]) / B[1,1] 
            U_new[0] = (c[0] - B[0,1] * U_new[1]) / B[0,0] 
        else: 
            U_new[0] = (c[0] - B[0,1] * U[1]) / B[0,0] 
            U_new[1] = (c[1] - B[1,0] * U_new[0]) / B[1,1] 
        
        epsilon = lin.norm(np.matmul(B,U_new) - c) 
        
    return U

In [7]:
eps    = 0.01
Beps   = .5* np.array([[eps+1/eps,eps-1/eps],[eps-1/eps,eps+1/eps]])
c      = np.array([1.,2.])
U_init = np.ones(2)
k_max  = 10**5
TOL    = .00001

U_sol = GS_altern(Beps, c, U_init, k_max, TOL)
print("Matrix\n", Beps,"\n")
print("Numerical solution\n", U_sol,"\n")
print("Residual\n",np.matmul(Beps,U_sol)-c)

Matrix
 [[ 50.005 -49.995]
 [-49.995  50.005]] 

Numerical solution
 [149.99449986 150.00449996] 

Residual
 [-1.00018224e-05  0.00000000e+00]


**c)** Compute the exact solution of this problem and compare it with these numerical solutions. Interprete the differences. 

*Hint: You may play with the different parameters of the numerical methods to analyze them.*

**Answer:** 

We compute the exact solution of this problem. 
\begin{align} 
BU = c 
\quad &\Longrightarrow \quad 
\begin{pmatrix} 
    50.005 & -49.995 \\ -49.995 & 50.005 
\end{pmatrix} 
\begin{pmatrix} 
    U_1 \\ U_2 
\end{pmatrix} = 
\begin{pmatrix} 
    1 \\ 2
\end{pmatrix} 
\\ 
\quad &\Longrightarrow \quad 
\begin{cases} 
    (50 + 0.005) U_1 - (50 - 0.005) U_2 = 1 \\ 
    - (50 - 0.005) U_1 + (50 + 0.005) U_2 = 2 
\end{cases} 
\end{align}

Adding both equations together, we obtain that 
$$ 
0.01 U_1 + 0.01 U_2 = 3 \quad \Longrightarrow \quad U_1 + U_2 = 300 
$$ 

Substituting $U_2 = 300 - U_1$, we obtain that 
$$ 
(50 + 0.005) U_1 - (50 - 0.005)(300 - U_1) = 1 
\quad \Longrightarrow \quad 
100 U_1 = 14999.5 
$$ 

Therefore, the exact solution is 
$$ 
\begin{cases} 
    U_1 = 149.995 \\ 
    U_2 = 150.005 
\end{cases} 
$$ 

By comparing the numerical solutions to the exact solutions, we see that none of the algorithm converges to the exact solutions due to a small $k_{max}$ value as well as a high tolerence value. Thus, by setting $k_{max} = 10^5$ instead of $100$, we see that all three algorithm converge to the exact value. Similarly, we can observe the same effect by setting $TOL = 0.00001$ instead of $0.001$. 